File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> freqMap = new HashMap<>();
        int left = 0, maxLength = 0, maxF = 0;
        char ch = '#';
        for(int right = 0; right<s.length(); right++){
            ch = s.charAt(right);
            freqMap.put(ch, freqMap.getOrDefault(ch,0) + 1);
            maxF = Math.max(maxF, freqMap.get(ch));
            while((right - left + 1) - maxF > k){
                ch = s.charAt(left);
                freqMap.put(ch, freqMap.get(ch) - 1);
                left++;
            }
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l, res = 0, 0
        freq = defaultdict(int)
        for r in range(len(s)):
            freq[s[r]] += 1
            while (r - l + 1) - max(freq.values()) > k:
                freq[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res
